### DISCUSSION TITLE
源码包的复用问题以及是否能在package内添加option？

### DISCUSSION THREAD

**Discussion by xqyjlj** [xmake-io/xmake#3062] [Q&A] [answered]
这里举一个例子，我现在程序引用了两个包，A包是一个实时操作系统，B包是一个flash芯片的驱动。在A包的源仓库，使用option+menu对其进行资源裁剪以及配置，例如芯片位宽，是否打开内存分配等。B包驱动兼容其公司旗下众多flash芯片，因此使用时也需要对其进行配置。

  现在问题来了，在使用package对这两个包进行管理的时候，由于package没有option这个组件，想要传递给package配置信息，也是只能进行手动传递 `add_requires("tbox", {configs = {small = true}})` ，现在这两个包，均是需要大量配置选项的，且比较灵活，如果手动对其进行配置的效率太低，而且容易出错，因此在这里是否能把包源码仓库的option继承给项目呢？

  除此之外也涉及到复用问题，即可能每个包在不同的项目中都会不一样，关联一下 #2888

**Comment by augustheart** [comment]
关注一下，我也正好想了解

**Comment by waruqi** [comment, marked as answer]
不能，你必须 package 里面通过 add_configs 一一对应映射一遍。。

参考

https://github.com/xmake-io/xmake-repo/blob/9c5059214aabe8aaac38e77b200af63150517b94/packages/f/ffmpeg/xmake.lua#L31-L43

然后自己的项目，options 自己定义传入 add_requires 的 configs。。

如果你想省去重复的 options 定义，可以吧所有共用的 options 放到单独的 options.lua 里面去维护，子包和当前项目包 都去 includes 进来复用就行了。。

**Reply by xqyjlj** [reply]
花了一晚上去研究怎么方便的去includes 复用的options.lua。现在options.lua存在包仓库中。那么在项目中怎么获取包仓库的options.lua路径呢？在包仓库的路径分别为`.xmake/cache/packages/2211/t/tbox/v1.7.1`，`.xmake/packages/t/tbox/v1.7.1/806b3ca1a99341c293542b7ffe40f7b8`，这些都不是一个固定的路径。在脚本域最早的on_load也晚于xmake f --menu的菜单显示啊

**Reply by waruqi** [reply]
目前不支持从包里面引入options，这个加载时机过早，你只能复制 options.lua但项目内，本地includes

目前只有 rule 可以走 package 分发，但也无法通过这个导出options

**Reply by xqyjlj** [reply]
有考虑增加这个功能吗，我可以帮忙开发并pr过来

**Reply by waruqi** [reply]
> 有考虑增加这个功能吗，我可以帮忙开发并pr过来

options 加载是过早的，packages 加载很滞后，我想支持都实现不了的，更别说其他人去 pr 实现了。。

而且，我也不打算支持。
